### PR TITLE
[ONNX] Fix for batchnorm training op mode

### PR DESCRIPTION
**Pull request by neginraoof**
Batchnorm should be exported in trainin/eval mode regardless of model's training mode.
This use-case is important for fine tuning models.

### CODE REVIEW

**Review by t10-13rocket — APPROVED** [review]

**Review by jiafatom — APPROVED** [review]
lgtm, please fix test errors.

File: torch/onnx/symbolic_opset12.py [hunk 1]
```diff
@@ -17,9 +17,9 @@ def einsum(g, equation, tensor_list):
 
 @parse_args('v', 'f', 'i')
 def dropout(g, input, p, train):
-    sym_help.assert_training_mode(train, "dropout")
+    sym_help.check_training_mode(train, "dropout")
     # in eval mode, dropout is non-op - if the node's train param is set to False, dropout is non-op
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset12.py:21]: update comment?

Comment (neginraoof) [reply torch/onnx/symbolic_opset12.py:21]: Do you mean the comment in line 21? I think this is not changed, since it is pointing to the training param for this node (op).

**Review by t10-13rocket — APPROVED** [review]

**Review by shubhambhokare1 — APPROVED** [review]
LGTM, thanks

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 2]
```diff
@@ -7372,11 +7480,63 @@ def forward(self, x):
...
+         zip(ort_outs1, [pt_outs1])]
+
         ort_sess2 = convert_to_onnx(model, input=(x,), opset_version=self.opset_version,
                                     training=torch.onnx.TrainingMode.EVAL)
         ort_outs2 = run_ort(ort_sess2, input=(x,))
-        [np.testing.assert_allclose(ort_out1, ort_out2, atol=1e-7, rtol=0.001) for ort_out1, ort_out2 in
-         zip(ort_outs1, ort_outs2)]
+        model.eval()
+        pt_outs2 = model(x)
+        [np.testing.assert_allclose(ort_out2, pt_out2.detach(), atol=1e-4, rtol=0.001) for ort_out2, pt_out2 in
```
Comment (fatcat-z) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:7492]: Change the format of rtol=0.001 to rtol=1e-3?

In addition, for atol, our normal testing uses 'atol=1e-7'. Should we keep the same one here?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 3]
```diff
@@ -7372,11 +7480,63 @@ def forward(self, x):
         ort_sess1 = convert_to_onnx(model, input=(x,), opset_version=self.opset_version,
                                     training=torch.onnx.TrainingMode.TRAINING)
         ort_outs1 = run_ort(ort_sess1, input=(x,))
+        pt_outs1 = model(x)
+        [np.testing.assert_allclose(ort_out1, pt_out1.detach(), atol=1e-4, rtol=0.001) for ort_out1, pt_out1 in
```
Comment (fatcat-z) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:7484]: Change the format of rtol=0.001 to rtol=1e-3?

In addition, for atol, our normal testing uses 'atol=1e-7'. Should we keep the same one here?

File: torch/onnx/symbolic_helper.py [hunk 4]
```diff
@@ -697,19 +697,23 @@ def _avgpool_helper(tuple_fn, padding, kernel_size, stride, divisor_override, na
...
         # if the model is a FuncModule. In this case we warn the user of
-        # the state and export depending on training_mode
+        # the state and export depending on op_mode
+        # This is to support use-cases of fixing certain layer weights
+        # in training.
         warnings.warn("ONNX export mode is set to " + training_mode +
                       " mode, but operator " + op_name + " is set to " +
-                      op_mode + " mode. The model will be exported in " +
-                      training_mode + ", as specified by the export mode.")
+                      op_mode + " mode. The operatos will be exported in " +
```
Comment (fatcat-z) [inline torch/onnx/symbolic_helper.py:714]: Typo: operatos -> operator

**Review by shubhambhokare1 — COMMENTED** [review]
Overall looks great and great test coverage, some minor comments (including Jay's)

File: test/onnx/test_operators.py [hunk 5]
```diff
@@ -681,7 +681,7 @@ def test_dropout_training(self):
 
     def test_dropout_opset12(self):
         x = torch.randn(3, 4, requires_grad=True)
-        self.assertONNX(lambda x: torch.max(functional.dropout(x)), x, opset_version=12)
+        self.assertONNX(lambda x: torch.max(functional.dropout(x, training=False)), x, opset_version=12)
```
Comment (BowenBao) [inline test/onnx/test_operators.py:684]: what is this change for?

File: torch/onnx/symbolic_helper.py [hunk 6]
```diff
@@ -700,19 +700,23 @@ def _avgpool_helper(tuple_fn, padding, kernel_size, stride, divisor_override, na
...
+        op_mode = "training " if op_train_mode else "inference"
         training_mode = "training " if _training_mode else "inference"
         # setting the model mode could result in op_mode != _training_mode
         # if the model is a FuncModule. In this case we warn the user of
-        # the state and export depending on training_mode
+        # the state and export depending on op_mode
+        # This is to support use-cases of fixing certain layer weights
+        # in training.
         warnings.warn("ONNX export mode is set to " + training_mode +
                       " mode, but operator " + op_name + " is set to " +
```
Comment (BowenBao) [inline torch/onnx/symbolic_helper.py:716]: Not to be done in this PR, we should update node naming to be consistent throughout IR graph and exported model, such that we can print which particular node this is in warning messages.

File: torch/onnx/symbolic_helper.py [hunk 7]
```diff
@@ -700,19 +700,23 @@ def _avgpool_helper(tuple_fn, padding, kernel_size, stride, divisor_override, na
...
     return padding
 
-def assert_training_mode(op_mode, op_name):
+
+def check_training_mode(op_train_mode, op_name):
     global _training_mode
-    op_mode = True if op_mode == 1 else False
-    if op_mode != _training_mode:
-        op_mode = "training " if op_mode else "inference"
+    op_train_mode = True if op_train_mode == 1 else False
```
Comment (BowenBao) [inline torch/onnx/symbolic_helper.py:706]: Checking my understanding, how is the op_mode created for operators in the first place? 
I assume users could set modes for ops explicitly. And without it the op will take the model mode. Is that correct?

File: torch/csrc/jit/passes/onnx/eval_peephole.cpp [hunk 8]
```diff
@@ -46,14 +46,18 @@ static void fuseConvBatchNorm(Block* b, ValueToParamPairMap& valsToParamsMap) {
...
-      if (bnNode->kind() != onnx::BatchNormalization) {
+
+      auto bnNode = origconvNode->output()->uses()[0].user;
+      if (bnNode->kind() != onnx::BatchNormalization)
         continue;
-      }
-      auto origconvNode = *it;
+      // Even when model is in eval mode, check if BatchNorm layer is in
+      // training mode
+      if (bnNode->outputs().size() != 1)
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/eval_peephole.cpp:58]: does batchnorm has a training attribute to check with, or is this only achievable by looking at output count?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 9]
```diff
@@ -7592,24 +7676,50 @@ def forward(self, x):
...
                                     training=torch.onnx.TrainingMode.EVAL)
         ort_outs2 = run_ort(ort_sess2, input=(x,))
-        [np.testing.assert_allclose(ort_out1, ort_out2, atol=1e-7, rtol=0.001) for ort_out1, ort_out2 in
-         zip(ort_outs1, ort_outs2)]
+        model.eval()
+        pt_outs2 = model(x)
+        [np.testing.assert_allclose(ort_out2, pt_out2.detach(), atol=1e-4, rtol=0.001) for ort_out2, pt_out2 in
+         zip(ort_outs2, [pt_outs2])]
+
+    @unittest.skip("Disable due to missing shape of Conv op output as BN input")
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:7691]: is this still active? why is it not resolved by shape inference?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 10]
```diff
@@ -7456,42 +7456,126 @@ def test_batchnorm_training(self):
...
 
-        model = MyModule()
-        x = torch.randn(10, 3, 128, 128)
+        x = torch.randn(10, 3, 64, 64) * 2
+        model_export = MyModule()
+        model_export.train()
+        outs = model_export(x)
+        pytorch_out = [out.detach().numpy() for out in outs]
+        ort_sess = convert_to_onnx(model_export, input=(x,), opset_version=self.opset_version, example_outputs=outs,
+                                   training=torch.onnx.TrainingMode.TRAINING,
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:7479]: why calling convert_to_onnx and compare results explicitly? Would adding training flag to `run_test` enough to resolve this?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 11]
```diff
@@ -7641,11 +7751,63 @@ def forward(self, x):
...
                                     training=torch.onnx.TrainingMode.EVAL)
         ort_outs2 = run_ort(ort_sess2, input=(x,))
-        [np.testing.assert_allclose(ort_out1, ort_out2, atol=1e-7, rtol=0.001) for ort_out1, ort_out2 in
-         zip(ort_outs1, ort_outs2)]
+        model.eval()
+        pt_outs2 = model(x)
+        [np.testing.assert_allclose(ort_out2, pt_out2.detach(), atol=1e-4, rtol=0.001) for ort_out2, pt_out2 in
+         zip(ort_outs2, [pt_outs2])]
+
+    @unittest.skip("Disable due to missing shape of Conv op output as BN input")
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:7766]: ditto

**Comment by BowenBao** [comment]
Looks good overall, have a few questions on some details. Also please check the CI failure.

File: test/onnx/test_operators.py (hunk 5 above)
Comment (neginraoof) [reply test/onnx/test_operators.py:684]: training is True by default

File: torch/csrc/jit/passes/onnx/eval_peephole.cpp (hunk 8 above)
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/eval_peephole.cpp:58]: Starting from opset 14, it has an attribute.

File: torch/onnx/symbolic_helper.py (hunk 7 above)
Comment (neginraoof) [reply torch/onnx/symbolic_helper.py:706]: Yes, see example in: test_batchnorm_eval_mode_train_layer

File: torch/onnx/symbolic_opset9.py [hunk 12]
```diff
@@ -1378,8 +1378,8 @@ def batch_norm(g, input, weight, bias, running_mean, running_var, training, mome
     out = g.op("BatchNormalization", input, weight, bias, running_mean, running_var,
                epsilon_f=eps,
                momentum_f=1 - momentum,
-               outputs=1 if not sym_help._training_mode else 5)
-    if not sym_help._training_mode:
+               outputs=1 if not training else 5)
```
Comment (shubhambhokare1) [inline torch/onnx/symbolic_opset9.py:1381]: Might need to replicate this for opset 14 BN

File: test/onnx/test_utility_funs.py [hunk 13]
```diff
@@ -917,21 +921,6 @@ def forward(self, x, y):
                                (TestCase,),
                                dict(TestUtilityFuns.__dict__, opset_version=13))
 
-# opset 11 tests
```
Comment (shubhambhokare1) [inline test/onnx/test_utility_funs.py:920]: Have removed this opset 14, might cause conflicts

**Review by shubhambhokare1 — COMMENTED** [review]
Overall looks good, could you look at batchnorm_helper added in opset 14 PR

File: torch/onnx/symbolic_opset9.py (hunk 12 above)
Comment (neginraoof) [reply torch/onnx/symbolic_opset9.py:1381]: Well this PR does not have opset 14 updates yet. I'll have to do that after rebase.

**Comment by neginraoof** [comment]
> Overall looks good, could you look at batchnorm_helper added in opset 14 PR

Well this PR does not have opset 14 updates yet. I'll have to do that after rebase.

File: torch/csrc/jit/passes/onnx/eval_peephole.cpp [hunk 14]
```diff
@@ -47,14 +47,17 @@ static void fuseConvBatchNorm(Block* b, ValueToParamPairMap& valsToParamsMap) {
       fuseConvBatchNorm(child_block, valsToParamsMap);
     }
     if (it->kind() == onnx::Conv) {
-      if (it->output()->uses().size() != 1) {
+      auto oldConv = *it;
+      auto bnNode = oldConv->outputs().at(0)->uses()[0].user;
```
Comment (fatcat-z) [inline torch/csrc/jit/passes/onnx/eval_peephole.cpp:51]: Does this change mean that we expect the Conv op will have more than 1 output, not like before anymore?
If so, how can we assure that the first output is BatchNorm? Should we go through all of existing outputs to find out if BatchNorm op exists?

Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/eval_peephole.cpp:51]: No, I don't expect Conv node to have more than one outputs.
The use of "at(i)" would throw a runtime error instead of a seg-fault in case of an issue when we mess up the nodes.

File: torch/csrc/jit/passes/onnx/eval_peephole.cpp [hunk 15]
```diff
@@ -47,14 +47,17 @@ static void fuseConvBatchNorm(Block* b, ValueToParamPairMap& valsToParamsMap) {
       fuseConvBatchNorm(child_block, valsToParamsMap);
     }
     if (it->kind() == onnx::Conv) {
-      if (it->output()->uses().size() != 1) {
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/eval_peephole.cpp:50]: We should need to check bn is the only user of this conv right?

Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/eval_peephole.cpp:50]: Just added this back. Thanks.

**Review by BowenBao — APPROVED** [review]
LGTM, thanks!
